File: py/day18.py
import time
import heapq
import math
from dataclasses import dataclass
from collections import deque, defaultdict, Counter
from typing import *

import numpy as np
import itertools
import re

import fire


def main(
        #infile: str='test_input.txt',
        infile: str = 'input.txt',
):
    print('hi!')

    lines = []
    with open(f'../inputs/day18/{infile}', 'r') as f:
        for i, line in enumerate(f.readlines()):
            lines.append(eval(line))

    print('=========')
    snums = [parse(l) for l in lines]
    t0 = time.time()
    x = snums[0]
    for i in range(1, len(snums)):
        x = Snum(x, snums[i])
        reduce(x)

    print('1:', magnitude(x), time.time() - t0, 's', x)

    t1 = time.time()
    max_mag = 0
    for i, j in itertools.product(range(len(snums)), range(len(snums))):
        x = Snum(parse(lines[i]), parse(lines[j]))
        reduce(x)
        mag = magnitude(x)
        if mag > max_mag:
            max_mag = mag

    print('2:', max_mag, time.time() - t1, 's')


@dataclass
class Snum:
    l: Optional['Snum'] = None
    r: Optional['Snum'] = None
    val: Optional[int] = None

    def is_leaf(self):
        return self.val is not None

    def add(self, val, dir):
        if self.is_leaf():
            self.val += val
            return

        if dir == 'l':
            self.l.add(val, dir)
        else:
            self.r.add(val, dir)

    def explode(self, depth=0):
        if self.is_leaf():
            return

        if depth == 4:
            lval = self.l.val
            rval = self.r.val
            self.l = None
            self.r = None
            self.val = 0
            return lval, rval
        x = self.l.explode(depth+1)
        if x:
            lval, rval = x
            if rval:
                self.r.add(rval, 'l')
            return lval, 0
        x = self.r.explode(depth+1)
        if x:
            lval, rval = x
            if lval:
                self.l.add(lval, 'r')
            return 0, rval

    def __str__(self):
        return str(self.val) if self.is_leaf() else f'[{str(self.l)},{str(self.r)}]'


def explode(x):
    depth = 0
    stack = list()
    cur = x
    prevprev, prev, rval = None, None, None
    exploded = False
    while cur or len(stack):
        if cur:
            stack.append((cur, depth))
            cur = cur.l
            depth += 1
        else:
            cur, depth = stack.pop()
            if cur.is_leaf():
                if rval is not None:
                    cur.val += rval
                    return True
                prevprev = prev
                prev = cur
            elif depth == 4:
                lval, rval = cur.l.val, cur.r.val
                cur.l, cur.r = None, None
                cur.val = 0
                exploded = True
                if prevprev:
                    prevprev.val += lval

            cur = cur.r
            depth += 1

    return exploded


def split(node):
    stack = list()
    cur = node
    while cur or len(stack):
        if cur:
            stack.append(cur)
            cur = cur.l
        else:
            cur = stack.pop()
            if cur.is_leaf() and cur.val > 9:
                cur.l = Snum(val=cur.val // 2)
                cur.r = Snum(val=(cur.val + 1) // 2)
                cur.val = None
                return True
            cur = cur.r

    return False


def magnitude(x):
    if x.is_leaf():
        return x.val
    else:
        return 3 * magnitude(x.l) + 2 * magnitude(x.r)


def reduce(x):
    while True:
        if explode(x):
            continue
        if split(x):
            continue
        break


def parse(x):
    if isinstance(x, list):
        return Snum(parse(x[0]), parse(x[1]))
    else:
        return Snum(val=x)


if __name__ == '__main__':
    fire.Fire(main)

File: go/day18.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"runtime/debug"
	"strconv"
	"time"
)

type Node struct {
	l *Node
	r *Node
	val int
	isLeaf bool
}

func (node *Node) toString() string {
	if node.isLeaf {
		return fmt.Sprintf("%v", node.val)
	} else {
		return fmt.Sprintf("[%v,%v]",
			node.l.toString(), 
			node.r.toString(),
		)
	}
}

func parse(snum string) *Node {
	if val, err := strconv.Atoi(snum); err == nil {
		return &Node{nil, nil, val, true}
	}
	// else parse out substrings
	var stack []int
	var comma int
	var left, right *Node
	for i, c := range snum {
		switch c {
		case '[': {
			stack = append(stack, i)
		}
		case ']': {
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				right = parse(snum[comma+1:i])
			}
		}
		case ',': {
			if len(stack) == 1 {
				left = parse(snum[1:i])
				comma = i
			}
		}
		}
	}

	return &Node{left, right, -1, false}
}

func explode(x *Node) bool {
	type Pair struct {
		node *Node
		depth int
	}
	cur := Pair{x, 0}
	var stack []Pair
	var prevprev, prev *Node
	var rval int
	found := false

	for cur.node != nil || len(stack) > 0 {
		if cur.node != nil {
			stack = append(stack, cur)
			cur = Pair{cur.node.l, cur.depth+1}
		} else {
			cur = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			// process cur
			if cur.node.isLeaf {
				if found {
					cur.node.val += rval
					return true
				}
				prevprev = prev
				prev = cur.node
			} else if cur.depth == 4 {
				if cur.node.isLeaf {
					log.Panicln(fmt.Sprintf("%v", cur.node.toString()))
				}
				lval := cur.node.l.val
				rval = cur.node.r.val
				found = true
				cur.node.l, cur.node.r = nil, nil
				cur.node.val, cur.node.isLeaf = 0, true
				if prevprev != nil {
					prevprev.val += lval
				}
			}
			cur = Pair{cur.node.r, cur.depth+1}
		}
	}
	return found
}

func split(cur *Node) bool {
	var stack []*Node
	for cur != nil || len(stack) > 0 {
		if cur != nil {
			stack = append(stack, cur)
			cur = cur.l
		} else {
			cur = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if cur.isLeaf && cur.val > 9 {
				cur.l = &Node{nil, nil, cur.val/2, true}
				cur.r = &Node{nil, nil, (cur.val+1)/2, true}
				cur.val = 0
				cur.isLeaf = false
				return true
			}
			cur = cur.r
		}
	}
	return false
}

func reduce(x *Node) {
	for true {
		if explode(x) {
			continue
		}
		if split(x) {
			continue
		}
		break
	}
}

func magnitude(x *Node) int {
	if x.isLeaf {
		return x.val
	} else {
		return 3*magnitude(x.l) + 2*magnitude(x.r)
	}
}


func main() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println(string(debug.Stack()))
		}
	}()

	var filename string = os.Args[1]
	t0 := time.Now()
	fmt.Printf("Reading from %s,", filename)
	file, err := os.Open(filename)
	if err != nil {
		log.Panicln(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		log.Panicln(err)
	}
	fmt.Printf(" took %v\n", time.Since(t0))

	t1 := time.Now()
	x := parse(lines[0])
	for _, line := range lines[1:] {
		node := parse(line)
		x = &Node{x, node, 0, false}
		reduce(x)
	}
	t := magnitude(x)
	fmt.Printf("1: %v, %v\n", t, time.Since(t1))

	t2 := time.Now()
	t = 0
	for i, line := range lines {
		for j, line2 := range lines {
			if i == j {
				continue
			}
			x = &Node{parse(line), parse(line2), 0, false}
			reduce(x)
			m := magnitude(x)
			if m > t {
				t = m
			}
		}
	}
	fmt.Printf("2: %v, %v\n", t, time.Since(t2))
}
